plugin/core/windows.py

```diff
@@ -214,6 +214,8 @@ def handle_did_open(self, view: ViewLike) -> None:
 
     def _notify_did_open(self, view: ViewLike, session: Session) -> None:
         language_id = self._view_language(view, session.config.name)
+        # There might be pending, no longer valid changes when restoring session.
+        self._pending_buffer_changes.pop(view.buffer_id(), None)
```

Comment:
This doesn't work when there is more than one session attached to the view.
The first initialized session will discard the changes here, which is OK.
The second initialized session will also discard the changes, which is not OK for the previously initialized session.

However, I'm OK with merging this. I'll add the proper fix in #1069.

Reply:
The loop that iterates session runs synchronously so it shouldn't be possible for any `on_text_changed` notifications to come during that iteration.

Reply:
The code in `notify_did_change` looks suspicious though as it will call `handle_did_open` when didOpen wasn't notified yet. So even though we'll clear the buffer, it will still be sent...

Reply:
Let's say the file content is `aaa`.
The user removes the last character, turning it into `aa`. Pending changes: "remove `a`".
Now session1 initialized, and it removes the pending changes. It sends `aa` as the initial document state. OK.
The user removes another character, turning the buffer into `a`. Pending changes: "remove `a`".
session2 initialized, and it removes the pending changes. Now, session1 will never get the pending change and will think the document is still `aa`.

Reply:
One possible fix is to forget about delayed/batched text changes and send them immediately (but only for the case that the server understands incremental text sync, of course).

Reply:
Right, I assume sessions can't be added dynamically like that.

Reply:
I'm okay with merging this anyway, but just be aware that there's still some subtleties left to consider.